### ISSUE TITLE
References to linting errors are in the position in the templated file, not in the reference to the raw file.

### ISSUE THREAD

**Issue by alanmcruickshank** [sqlfluff/sqlfluff#273]
For example if we have a file such as:

```
select
    {% for elem in ('a', 'b', 'c' , 'd') %}
    {{elem}} as col_{{elem}},
    {% endfor %}
    some_other_col
from my_table AS foo
```

And lint for something that will happen late in the file:

```
$ sqlfluff lint test.sql --rules L010
== [test.sql] FAIL
L:  12 | P:  15 | L010 | Inconsistent capitalisation of keywords.
```

The linting error shows as being on line 12, which is true in the templated file, but *not* in the raw file, where the issue is with the `AS` keyword on line 6. This makes it much harder to track down exactly where the error is in the original file.

My suggestion is that while the current behavior should still be an option, that the default behavior should be to report the position in the *raw* file, i,e. that the error shown above should be reported at `L:  5 | P: 15`.

**Comment by alanmcruickshank** [comment]
Update:

Conversation on the pallets projects discord:

> @amshank In having a similar requirement myself, but for debugging, I ended up having to write my own template engine.  :confused: https://stackoverflow.com/a/54678540/211827 is a question I asked long ago about how to efficiently package the line association information, ended up answering myself (RLE), includes sample approach to dynamically reversing the generated line number back to the source line that triggered it from the compressed data.

>Trick is to get that data generated in the first place…
> @GothAlice - thanks for the pointers. I think I can get at the data using Template.generate() , which will at least separate bits out and then I'll see if I can use something similar to your approach to get from there back to the components of the template. :thumbsup:
If anyone else has any bright ideas - I'm all ears :slight_smile: - otherwise I'll go with an approach similar to the one above.

Looks like we're going to have to do something hacky for this. Outline scope:

- `LintedFile` provides an interface for converting reference positions from _templated space_ to _raw space_.
- Position markers have a property to indicate which _space_ they're in. Defaulting to _templated space_ because that's how most of the code works so far.
- Templaters provide the meat of how to do this (probably as an additional output of the templating step), and for the jinja templater it's done using the `Template.generate()` method to iteratively generate strings from the template, and then matching them against the tokens from `Environment.lex()` so do a _best effort_ approximation of the location in the original file.
- Configuration defaults to _raw space_ as the way of converting references.
- (some what seperately) we use `jinja2.meta.find_undeclared_variables` to generate better references to any missing jinja variables, _which would be provided in "raw space" already_.

**Comment by alanmcruickshank** [comment]
I'd love to add this to the roadmap for 0.4.0, but I'm a bit out of ideas on how to achieve it reliably. Any ideas?

**Comment by dorzey** [comment]
Does this issue also explain the mismatch I see on my DBT files?

Could we build a sourcemap?

**Comment by alanmcruickshank** [comment]
> Does this issue also explain the mismatch I see on my DBT files?

Almost certainly.

> Could we build a sourcemap?

Could you elaborate? I'm open to all ideas here so if the idea works then I'm game.

**Comment by dorzey** [comment]
>> Could we build a sourcemap?

> Could you elaborate? I'm open to all ideas here so if the idea works then I'm game.

Sourcemaps allow you to map compressed/minified code back to it's original position in the source file; so for example I can debug JavaScript in the browser even on compressed/minified files.

This feels like the inverse problem.

I should caveat that I'm not sure how to implement one or if such an approach would fit into the sqlfluff architecture.

**Comment by alanmcruickshank** [comment]
Heads up team that I'm working on this one at the moment. I've sunk quite a lot of time into this so don't want anyone else to pick it up in the mean time. Should have sometime to review in the next few days.
